### PR TITLE
S3 backing store

### LINKED ISSUE #116: Mediachain backup in S3
In order to protect from disasters we should back up the mediachain in S3.
This can follow block generation: every time a new block is generated, a supporting node would aggregate the block together with all the new objects pointed by the block together into a fat block and store it in S3.

This allows us to recover from disasters while losing at most one block of updates.
It also simplifies adding IPFS pinning nodes that provide the whole mediachain, as such nodes can use the fat blocks directly from S3.

### PR DESCRIPTION
**Pull request by vyzo**
close #116 
Adds support for backing up and restoring the mediachain in S3:
- S3BackingStore implements the backup process as a transactor client which archives fat blocks into S3
- S3Restore implements restore by replaying the blockchain from S3 backed up fat blocks.

### CODE REVIEW

File: protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala [hunk 1]
```diff
@@ -601,6 +606,45 @@ object CborSerialization {
...
+          block = block,
+          data = data
+        )
+    
+    def dataFromCbor(records: CArray) = {
+      def loop(rest: List[CValue], map: Map[Reference, Array[Byte]])
+      : Xor[DeserializationError, Map[Reference, Array[Byte]]] = {
+        rest match {
+          case CArray(List(refCbor: CMap, data: CBytes)) :: rest =>
+            ReferenceDeserializer.fromCMap(refCbor) match {
```
Comment (parkan) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:633]: I feel like this would be easier to read as a flatMap, since you're not doing anything with the other branch and it'll short-circuit anyway

File: protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala [hunk 2]
```diff
@@ -601,6 +606,45 @@ object CborSerialization {
...
+              case Xor.Right(ref) =>
+                loop(rest, map + (ref -> data.bytes))
```
Comment (parkan) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:635]: I'm kind of surprised you didn't `reverse` the list and build this through normal returns from the tail recursion

Comment (bigs) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:635]: :+1:

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  client.addStateListener(this)
+
+  def connect(addrs: List[String]) {
+    cluster = Some(addrs)
+    client.connect(addrs)
+    client.listen(this)
+  }
+  
+  def close() {
+    exec.shutdownNow()
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:42]: do we really need to aggressively shut down everything active?

File: protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala (hunk 1 above)
Comment (vyzo) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:633]: flat mapping will break tail recursion, as it won't be able to do a goto from the closure inside flatMap.

File: protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala (hunk 2 above)
Comment (vyzo) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:635]: It feels more natural to me to do a folding loop using an accumulator; kind of a scheme idiom.
Also, the order of iteration is not important, so the reverse doesn't add anything.

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala (hunk 3 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:42]: It is kind of a safe operation -- a restart will pick up where you stopped and you can't have incomplete puts in S3.
Also,  if you are calling close here you are probably exiting the entire program and you won't it to stop asap.

Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:42]: I'll change it to plain shutdown to let any actual writes in progress a chance to complete without an InterruptedException.

Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:42]: Actually plain shutdown is not quite correct, because we proceed to close the client used by the background threads anyway. Better to immediately shutdown with interrupt.
Will revert the change.

Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:42]: hmm ok maybe let's revisit this later

File: protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala [hunk 4]
```diff
@@ -601,6 +606,47 @@ object CborSerialization {
...
+        data <- dataFromCbor(xdata)
+      } yield
+        JournalBlockArchive(
+          ref = ref,
+          block = block,
+          data = data
+        )
+    
+    def dataFromCbor(records: CArray) = {
+      @tailrec def loop(rest: List[CValue], map: Map[Reference, Array[Byte]])
```
Comment (parkan) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:629]: hmm you're right using flatMap breaks the @tailrec, I'd have assumed that the two statements would get compiled to equivalent bytecode :\

Comment (parkan) [inline protocol/src/main/scala/io/mediachain/protocol/CborSerialization.scala:629]: my apologies to the user named "tailrec", they must get this a lot

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    logger.info("Fetching current block")
+    val op = client.currentBlock
+    val block = Await.result(op, Duration.Inf)
+    logger.info("Current blockchain pointer is " + block.chain)
+    block.chain.foreach { execWriteBlock(_) }
+  }
+  
+  private def execWriteBlock(ref: Reference) {
+    this.synchronized {
+      if (!blocks.contains(ref)) {
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:92]: am I reading this wrong, or will this back up the _entire_ chain every time?

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+  
+  // JournalListener
+  def onJournalCommit(entry: JournalEntry) {}
+
+  def onJournalBlock(ref: Reference) {
+    logger.info(s"New block ${ref}")
+    execWriteBlock(ref)
+  }
+  
+  private def withErrorLog(expr: => Unit) {
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:73]: I was about to say "why don't we just collect the futures and attach an onFailure handler" but then I went to read about java Futures and got really sad.

I kind of want to just rewrite everything like

```
implicit val execContext = ExecutionContext.fromExecutor(exec)
def logError = PartialFunction[Exception, Unit] = { case e: Exception => ... }
Future { writeCurrentBlock() }.recover(logError)
...
```

(maybe `Future { blocking....`)

since s3 client doesn't even use java futures, right?

File: transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  import S3BackingStore.multihashRef
+  private val logger = LoggerFactory.getLogger(classOf[S3Restore])
+  private val datastore = new DynamoDatastore(config.dynamo)
+  private val s3 = new AmazonS3Client()
+  private val s3bucket = config.s3bucket
+  private val client = Client.build(config.sslConfig)
+  // dynamo background writing
+  @volatile private var shutdown = false
+  private val queue = new LinkedBlockingQueue[(Reference, Array[Byte])]
+  private val writers = (1 to Runtime.getRuntime.availableProcessors).map { n =>
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala:31]: decided to manually manage these instead of using a pool?

File: transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    logger.info(s"Restoring ${cluster} from ${key}")
+    client.connect(cluster)
+    doRestore(key)
+    close()
+  }
+  
+  def close() {
+    logger.info("Shutting down...")
+    shutdown = true
+    writers.foreach(_.join())
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala:50]: don't need to do this w/pool

**Comment by parkan** [comment]
A lot of stuff happens here, for the most part it looks OK (other than huge scope for one PR!). Not super into the java-style error and thread handling but I get that it's kind of informed by using the s3 libs etc that don't even use futures of any kind.

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala (hunk 5 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:92]: It will walk up the chain until it finds a block it already has in the list of visited blocks or reaches nil.
It will not write a block a second time, because it will either be in the visited blocks set (which will make execWriteBlock a noop) or already in S3 (the write checks to see if it is already present in S3 with listObjects)

The reason for this implementation is that it allows the backing store to catch up in the block chain and be easily restartable and tolerant to transient client failures.

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala (hunk 6 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:73]: s3 interface is synchronous (as are all service interfaces in AWS sdk)

File: transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala (hunk 7 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala:31]: yeah, I find it cleaner when the threads are specialized.

**Comment by vyzo** [comment]
Yeah, we kind of have to write synchronous java-style code when using the aws sdk.

**Comment by vyzo** [comment]
I think we want to add backoff/retry around S3 errors as well - the one thing currently missing.

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala (hunk 5 above)
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:92]: ah ok the listObjects call inside the write was what I was missing (concern was that the set wouldn't be populated on a repeated run)

File: transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala (hunk 7 above)
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala:31]: the "preferred" java way to do that is

```
namedThreadFactory = new ThreadFactoryBuilder()
                                       .setNameFormat("S3Restore@${this.hashCode}#write-%d").build()
```

and pass it to executor

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala (hunk 6 above)
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:73]: right, so you would wrap them like this

``` scala
def putObject(s3bucket: String, key: String, gzpath: String): Future = Future {
  blocking {
    s3.putObject(s3bucket, key, new File(gzpath)
  }
}
```

this gives you a nice future-based interface and also some intelligent thread creation if you use ForkJoinPool

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+    } catch {
+      case e: Throwable =>
+        logger.error("Unhandled exception in task", e)
+    }
+  }
+  
+  private def writeCurrentBlock() {
+    logger.info("Fetching current block")
+    val op = client.currentBlock
+    val block = Await.result(op, Duration.Inf)
```
Comment (parkan) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:85]: is this Inf for a good reason?

**Comment by parkan** [comment]
yeah agreed on backoff-retry

File: transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala (hunk 9 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala:85]: there is no reasonable timeout here -- and the logic would have to retry it anyway, with multiple pending currentBlocks.

File: transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala (hunk 7 above)
Comment (vyzo) [inline transactor/src/main/scala/io/mediachain/copycat/S3Restore.scala]: ugh -- a builder for the factory!
